dhis-2/dhis-services/dhis-service-validation/src/test/java/org/hisp/dhis/outlierdetection/service/OutlierDetectionServiceMinMaxTest.java

```diff
@@ -0,0 +1,239 @@
+
+        idObjectManager.save( deA );
+        idObjectManager.save( deB );
+
+        ouA = createOrganisationUnit( 'A' );
+        ouB = createOrganisationUnit( 'B' );
+
+        idObjectManager.save( ouA );
+        idObjectManager.save( ouB );
+
+        coc = categoryService.getDefaultCategoryOptionCombo();
+    }
+
+    @Test
+    public void testGetFromQuery()
+    {
+        OutlierDetectionQuery query = new OutlierDetectionQuery();
+        query.setDe( Lists.newArrayList( "deabcdefghA", "deabcdefghB" ) );
+        query.setStartDate( getDate( 2020, 1, 1 ) );
+        query.setEndDate( getDate( 2020, 6, 1 ) );
+        query.setOu( Lists.newArrayList( "ouabcdefghA", "ouabcdefghB" ) );
+        query.setAlgorithm( OutlierDetectionAlgorithm.MIN_MAX );
+        query.setMaxResults( 200 );
+
+        OutlierDetectionRequest request = subject.getFromQuery( query );
+
+        assertEquals( 2, request.getDataElements().size() );
+        assertEquals( 2, request.getOrgUnits().size() );
+        assertEquals( getDate( 2020, 1, 1 ), request.getStartDate() );
+        assertEquals( getDate( 2020, 6, 1 ), request.getEndDate() );
+        assertEquals( OutlierDetectionAlgorithm.MIN_MAX, request.getAlgorithm() );
+        assertEquals( 200, request.getMaxResults() );
+    }
+
+    @Test
+    public void testGetOutlierValues()
+    {
+        addMinMaxValues(
+            new MinMaxDataElement( deA, ouA, coc, 40, 60 ),
+            new MinMaxDataElement( deB, ouA, coc, 45, 65 ) );
+
+        // 34, 39, 68, 91, 42, 45, 68, 87 are outlier values outside the min-max range
+
+        addDataValues(
+            new DataValue( deA, m01, ouA, coc, coc, "50" ), new DataValue( deA, m07, ouA, coc, coc, "51" ),
+            new DataValue( deA, m02, ouA, coc, coc, "34" ), new DataValue( deA, m08, ouA, coc, coc, "59" ),
+            new DataValue( deA, m03, ouA, coc, coc, "58" ), new DataValue( deA, m09, ouA, coc, coc, "39" ),
+            new DataValue( deA, m04, ouA, coc, coc, "68" ), new DataValue( deA, m10, ouA, coc, coc, "52" ),
+            new DataValue( deA, m05, ouA, coc, coc, "51" ), new DataValue( deA, m11, ouA, coc, coc, "58" ),
+            new DataValue( deA, m06, ouA, coc, coc, "40" ), new DataValue( deA, m12, ouA, coc, coc, "91" ),
+
+            new DataValue( deB, m01, ouA, coc, coc, "42" ), new DataValue( deB, m02, ouA, coc, coc, "48" ),
+            new DataValue( deB, m03, ouA, coc, coc, "45" ), new DataValue( deB, m04, ouA, coc, coc, "46" ),
+            new DataValue( deB, m05, ouA, coc, coc, "49" ), new DataValue( deB, m06, ouA, coc, coc, "68" ),
+            new DataValue( deB, m07, ouA, coc, coc, "48" ), new DataValue( deB, m08, ouA, coc, coc, "49" ),
+            new DataValue( deB, m09, ouA, coc, coc, "52" ), new DataValue( deB, m10, ouA, coc, coc, "47" ),
+            new DataValue( deB, m11, ouA, coc, coc, "11" ), new DataValue( deB, m12, ouA, coc, coc, "87" ) );
```

Comment:
I can see how this gives a good overview on one hand but it is also repetitive and makes it hard to see the data that matters to us in the test. I think a 2 loops for `deA` and `deB` and an array of just the relevant numbers would make this more clear.

Reply:
Right. Was considering that but I would argue it is easier to read the way it is now. E.g. if we want to add more org units later it becomes yet another level in the loop and I would say readability will suffer.